app/core/store/modules/levelSessions.js

```diff
@@ -0,0 +1,118 @@
+import levelSessionsApi from 'core/api/level-sessions'
+
+const SESSIONS_PER_REQUEST = 10
+
+export default {
+  namespaced: true,
+
+  state: {
+    loading: {
+      sessionsByClassroom: {}
+    },
+
+    // CLASSROOM_ID: {
+    //   sessions: [],
+    //   levelCompletionsByUser: {
+    //      LEVEL_ID: true / false
+    //    }
+    // }
+    levelSessionsByClassroom: {}
+  },
+
+  mutations: {
+    toggleClassroomLoading: (state, classroomId) => {
+      let loading = true;
+      if (state.loading.sessionsByClassroom[classroomId]) {
+        loading = false
+      }
+
+      Vue.set(state.loading.sessionsByClassroom, classroomId, loading)
+    },
+
+    initSessionsByClassroomState: (state, classroomId) => {
+      if (state.levelSessionsByClassroom[classroomId]) {
+        return
+      }
+
+      Vue.set(state.levelSessionsByClassroom, classroomId, {
+        sessions: [],
+        levelCompletionsByUser: {}
+      })
+    },
+
+    setSessionsForClassroom: (state, { classroomId, sessions }) => {
+      state.levelSessionsByClassroom[classroomId].sessions = sessions
+      state.levelSessionsByClassroom[classroomId].levelCompletionsByUser = {}
+    },
+
+    addLevelCompletionsByUserForClassroom: (state, { classroomId, levelCompletionsByUser }) => {
+      state.levelSessionsByClassroom[classroomId].levelCompletionsByUser = levelCompletionsByUser
+    }
+  },
+
+  // TODO add a way to clear out old level session data
+  actions: {
+    // TODO how do we handle two sets of parallel requests here? (ie user vists page, hits back and then visits again quickly)
+    fetchForClassroomMembers: ({ commit }, classroom) => {
+      commit('toggleClassroomLoading', classroom._id)
+      commit('initSessionsByClassroomState', classroom._id)
+
+      // TODO comment what next line is doing
+      let requests = Array.from(
+        Array(parseInt(classroom.members.length / SESSIONS_PER_REQUEST, 10) + 1)
+      )
+
+      requests = requests.map((v, i) =>
+        levelSessionsApi.fetchForClassroomMembers(classroom._id, {
+          data: {
+            memberLimit: SESSIONS_PER_REQUEST,
+            memberSkip: i * SESSIONS_PER_REQUEST,
+            project: 'state.complete,level,creator,changed,created,dateFirstCompleted,submitted,codeConcepts'
+          }
+        })
+      )
+
+      return Promise.all(requests)
```

Comment:
A single request failing would reject the whole promise. Is this likely, especially for much larger classes.

This may also benefit from client side throttling. If this action is turned into an async method you can yield over the promises handling each individually. If this is too slow you could also await smaller batches of levelsessions.

This is an optimization that isn't required now.

Reply:
Great points here 👍 

With the background stats generator I'd like to try to move this to the server.  I think the query will be very similar to the projects query so I'll attempt to knock both out at the same time.  If that is successful this will go away :)

I added a note to https://app.asana.com/0/1115428071773315/1118364384628023/f to follow up on this